File: app/src/main/java/com/returnnotfound/stain/utils/FileUtils.java
package com.returnnotfound.stain.utils;

import android.annotation.SuppressLint;
import android.content.ContentUris;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.provider.DocumentsContract;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.webkit.MimeTypeMap;

import java.io.File;

import static com.returnnotfound.stain.CoreDefaultKt.MIME_TYPE_IMAGE_JPEG;

public abstract class FileUtils {

  @SuppressLint("NewApi")
  public static String getPathFromUri(final Context context, final Uri uri) {
    final boolean isKitKat = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT;
    // DocumentProvider
    if (isKitKat && DocumentsContract.isDocumentUri(context, uri)) {
      // ExternalStorageProvider
      if (isExternalStorageDocument(uri)) {
        final String docId = DocumentsContract.getDocumentId(uri);
        final String[] split = docId.split(":");
        final String type = split[0];

        if ("primary".equalsIgnoreCase(type)) {
          return Environment.getExternalStorageDirectory() + "/" + split[1];
        }
      }
      // DownloadsProvider
      else if (isDownloadsDocument(uri)) {
        Cursor cursor = null;
        try {
          cursor = context.getContentResolver().query(uri, new String[]{MediaStore.MediaColumns.DISPLAY_NAME}, null, null, null);
          if (cursor != null && cursor.moveToFirst()) {
            String fileName = cursor.getString(0);
            String path = Environment.getExternalStorageDirectory().toString() + "/Download/" + fileName;
            if (!TextUtils.isEmpty(path)) {
              return path;
            }
          }
        } finally {
          if (cursor != null)
            cursor.close();
        }
        final String id = DocumentsContract.getDocumentId(uri);
        try {
          final Uri contentUri = ContentUris.withAppendedId(
              Uri.parse("content://downloads/public_downloads"), Long.valueOf(id));

          return getDataColumn(context, contentUri, null, null);
        } catch (NumberFormatException e) {
          //In Android 8 and Android P the id is not a number
          return uri.getPath().replaceFirst("^/document/raw:", "").replaceFirst("^raw:", "");
        }
      }
      // MediaProvider
      else if (isMediaDocument(uri)) {
        final String docId = DocumentsContract.getDocumentId(uri);
        final String[] split = docId.split(":");
        final String type = split[0];

        Uri contentUri = null;
        if ("image".equals(type)) {
          contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
        } else if ("video".equals(type)) {
          contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
        } else if ("audio".equals(type)) {
          contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
        }

        final String selection = "_id=?";
        final String[] selectionArgs = new String[]{
            split[1]
        };

        return getDataColumn(context, contentUri, selection, selectionArgs);
      }
    }
    // MediaStore (and general)
    else if ("content".equalsIgnoreCase(uri.getScheme())) {

      // Return the remote address
      if (isGooglePhotosUri(uri))
        return uri.getLastPathSegment();

      return getDataColumn(context, uri, null, null);
    }
    // File
    else if ("file".equalsIgnoreCase(uri.getScheme())) {
      return uri.getPath();
    }

    return null;
  }

  public static String getDataColumn(Context context, Uri uri, String selection, String[] selectionArgs) {

    Cursor cursor = null;
    final String column = "_data";
    final String[] projection = {
        column
    };

    try {
      cursor = context.getContentResolver().query(uri, projection, selection, selectionArgs,
          null);
      if (cursor != null && cursor.moveToFirst()) {
        final int index = cursor.getColumnIndexOrThrow(column);
        return cursor.getString(index);
      }
    } finally {
      if (cursor != null)
        cursor.close();
    }
    return null;
  }

  public static boolean isExternalStorageDocument(Uri uri) {
    return "com.android.externalstorage.documents".equals(uri.getAuthority());
  }

  public static boolean isDownloadsDocument(Uri uri) {
    return "com.android.providers.downloads.documents".equals(uri.getAuthority());
  }

  public static boolean isMediaDocument(Uri uri) {
    return "com.android.providers.media.documents".equals(uri.getAuthority());
  }

  public static boolean isGooglePhotosUri(Uri uri) {
    return "com.google.android.apps.photos.content".equals(uri.getAuthority());
  }

  public static String getMimeType(File file) {
    String path = file.getPath();
    path = path.replace(" ", "");
    String extension = MimeTypeMap.getFileExtensionFromUrl(path).toLowerCase();
    String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
    if (StringUtils.isNullOrEmpty(mimeType)) {
      mimeType = MIME_TYPE_IMAGE_JPEG;
    }
    return mimeType;
  }

  public static long getSize(String path) {
    File file = new File(path);
    return file.length();
  }

  public static long getSizeMB(String path) {
    File file = new File(path);
    return file.length() / 1024 / 1024;
  }
}

File: app/src/main/java/com/returnnotfound/stain/base/utils/FileUtils.kt
package com.returnnotfound.stain.base.utils

import android.annotation.SuppressLint
import android.content.ContentUris
import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.DocumentsContract
import android.provider.MediaStore
import android.text.TextUtils
import android.webkit.MimeTypeMap
import com.returnnotfound.stain.base.MIME_TYPE_IMAGE_JPEG
import java.io.File

object FileUtils {
  @SuppressLint("NewApi")
  @JvmStatic
  fun getPathFromUri(context: Context, uri: Uri): String? {
    val isKitKat = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT
    // DocumentProvider
    if (isKitKat && DocumentsContract.isDocumentUri(context, uri)) {
      // ExternalStorageProvider
      if (isExternalStorageDocument(uri)) {
        val docId = DocumentsContract.getDocumentId(uri)
        val split = docId.split(":").toTypedArray()
        val type = split[0]
        if ("primary".equals(type, ignoreCase = true)) {
          return Environment.getExternalStorageDirectory()
            .toString() + "/" + split[1]
        }
      } else if (isDownloadsDocument(uri)) {
        var cursor: Cursor? = null
        try {
          cursor = context.contentResolver.query(
            uri,
            arrayOf(MediaStore.MediaColumns.DISPLAY_NAME),
            null,
            null,
            null
          )
          if (cursor != null && cursor.moveToFirst()) {
            val fileName = cursor.getString(0)
            val path = Environment.getExternalStorageDirectory()
              .toString() + "/Download/" + fileName
            if (!TextUtils.isEmpty(path)) {
              return path
            }
          }
        } finally {
          cursor?.close()
        }
        val id = DocumentsContract.getDocumentId(uri)
        return try {
          val contentUri = ContentUris.withAppendedId(
            Uri.parse("content://downloads/public_downloads"),
            java.lang.Long.valueOf(id)
          )
          getDataColumn(
            context,
            contentUri,
            null,
            null
          )
        } catch (e: NumberFormatException) {
          //In Android 8 and Android P the id is not a number
          uri.path!!.replaceFirst("^/document/raw:".toRegex(), "")
            .replaceFirst("^raw:".toRegex(), "")
        }
      } else if (isMediaDocument(uri)) {
        val docId = DocumentsContract.getDocumentId(uri)
        val split = docId.split(":").toTypedArray()
        val type = split[0]
        var contentUri: Uri? = null
        if ("image" == type) {
          contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI
        } else if ("video" == type) {
          contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI
        } else if ("audio" == type) {
          contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
        }
        val selection = "_id=?"
        val selectionArgs = arrayOf(
          split[1]
        )
        return getDataColumn(
          context,
          contentUri,
          selection,
          selectionArgs
        )
      }
    } else if ("content".equals(uri.scheme, ignoreCase = true)) {

      // Return the remote address
      return if (isGooglePhotosUri(uri)) uri.lastPathSegment else getDataColumn(
        context,
        uri,
        null,
        null
      )
    } else if ("file".equals(uri.scheme, ignoreCase = true)) {
      return uri.path
    }
    return null
  }

  @JvmStatic
  fun getDataColumn(
    context: Context,
    uri: Uri?,
    selection: String?,
    selectionArgs: Array<String>?
  ): String? {
    var cursor: Cursor? = null
    val column = "_data"
    val projection = arrayOf(
      column
    )
    try {
      cursor = context.contentResolver.query(
        uri!!, projection, selection, selectionArgs,
        null
      )
      if (cursor != null && cursor.moveToFirst()) {
        val index = cursor.getColumnIndexOrThrow(column)
        return cursor.getString(index)
      }
    } finally {
      cursor?.close()
    }
    return null
  }

  @JvmStatic
  fun isExternalStorageDocument(uri: Uri): Boolean {
    return "com.android.externalstorage.documents" == uri.authority
  }

  @JvmStatic
  fun isDownloadsDocument(uri: Uri): Boolean {
    return "com.android.providers.downloads.documents" == uri.authority
  }

  @JvmStatic
  fun isMediaDocument(uri: Uri): Boolean {
    return "com.android.providers.media.documents" == uri.authority
  }

  @JvmStatic
  fun isGooglePhotosUri(uri: Uri): Boolean {
    return "com.google.android.apps.photos.content" == uri.authority
  }

  @JvmStatic
  fun getMimeType(file: File): String? {
    var path = file.path
    path = path.replace(" ", "")
    val extension = MimeTypeMap.getFileExtensionFromUrl(path).toLowerCase()
    var mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension)
    if (StringUtils.isNullOrEmpty(mimeType)) {
      mimeType = MIME_TYPE_IMAGE_JPEG
    }
    return mimeType
  }

  @JvmStatic
  fun getSize(path: String?): Long {
    val file = File(path)
    return file.length()
  }

  @JvmStatic
  fun getSizeMB(path: String?): Long {
    val file = File(path)
    return file.length() / 1024 / 1024
  }
}